torch/jit/_script.py

```diff
@@ -1065,11 +1079,95 @@ def create_script_list(obj, type_hint=None):
     return torch._C.ScriptList(obj)  # type: ignore[attr-defined]
 
 
+# Defined in torch/csrc/jit/python/script_init.cpp
+ResolutionCallback = Callable[[str], Callable[..., Any]]
+ClassVar = TypeVar("ClassVar", bound=type)
+P = ParamSpec("P")
+R = TypeVar("R", covariant=True)  # return value
+
+
+@overload
+def script(
+    obj: Type[Module],
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> Never:
+    ...
+
+
+@overload
+def script(
+    obj: Dict,
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> torch.ScriptDict:
+    ...
+
+
+@overload
+def script(
+    obj: List,
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> torch.ScriptList:
+    ...
+
+
+@overload
+def script(  # type: ignore[misc]
+    obj: Module,
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> RecursiveScriptModule:
+    ...
+
+
+@overload
+def script(  # type: ignore[misc]
+    obj: ClassVar,
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> ClassVar:
+    ...
+
+
+@overload
+def script(  # type: ignore[misc]
+    obj: Callable[P, R],
+    optimize: Optional[bool] = None,
+    _frames_up: int = 0,
+    _rcb: Optional[ResolutionCallback] = None,
+    example_inputs: Union[List[Tuple], Dict[Callable, List[Tuple]], None] = None,
+) -> ScriptFunction[P, R]:
+    ...
```

Comment:
Added ParamSpec which ensures `mypy` recognizes that jit.script doesn't modify the signature of the callable.